src/circuit/gadget/poseidon/pow5.rs

```diff
@@ -399,17 +386,19 @@ impl<F: FieldExt, S: Spec<F, WIDTH, 2>> PoseidonDuplexInstructions<F, S, WIDTH,
                     Ok(StateWord { var, value })
                 };
 
-                Ok([
-                    constrain_output_word(0)?,
-                    constrain_output_word(1)?,
-                    constrain_output_word(2)?,
-                ])
+                let output: Result<Vec<_>, Error> = (0..WIDTH).map(constrain_output_word).collect();
+                output.map(|output| output.try_into().unwrap())
             },
         )
     }
 
-    fn get_output(state: &State<Self::Word, WIDTH>) -> SpongeState<Self::Word, 2> {
-        [Some(state[0]), Some(state[1])]
+    fn get_output(state: &State<Self::Word, WIDTH>) -> SpongeState<Self::Word, RATE> {
+        state[..RATE]
+            .iter()
+            .map(|word| Some(*word))
+            .collect::<Vec<_>>()
+            .try_into()
+            .unwrap()
```

Comment:
`array::map`, which is [stable in Rust 1.55](https://github.com/rust-lang/rust/commit/7c5cabe30f4c6402e06ab0240cdfc066e7c70e8e#diff-534d866e6adad8a1f7577c2f1b3eed0b05d5f001c0d5032ac226393736940381R312), would be so much nicer!

librustzcash has a [rust-toolchain](https://github.com/zcash/librustzcash/blob/master/rust-toolchain) for Rust 1.51, but zcashd [pins Rust 1.56.1](https://github.com/zcash/zcash/blob/master/depends/packages/native_rust.mk#L2).

Reply:
Yep, I plan to migrate everything to that at some point, but 1.51 is a reasonable schelling point for now, matching most of our dependencies. I think 1.56 will be the next MSRV bump (matching what it looks like the RustCrypto ecosystem will be moving to).